packages/harmony/src/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,178 @@
+import { forwardRef, useId } from 'react'
+
+import cn from 'classnames'
+
+import { Text, TextSize } from 'components/typography'
+
+import { Flex } from '../../layout'
+
+import { HelperText } from './HelperText'
+import styles from './TextInput.module.css'
+import { TextInputSize, type TextInputProps } from './types'
+import { useFocusState } from './useFocusState'
+
+const TextInput = forwardRef<HTMLInputElement, TextInputProps>((props, ref) => {
+  const {
+    required,
+    className,
+    inputRootClassName,
+    maxLength,
+    showMaxLengthThreshold = 0.7,
+    maxLengthWarningThreshold = 0.9,
+    size = TextInputSize.DEFAULT,
+    hideLabel,
+    label: labelProp,
+    value,
+    id: idProp,
+    children,
+    warning: warningProp,
+    error,
+    className: inputClassName,
+    disabled,
+    onFocus: onFocusProp,
+    onBlur: onBlurProp,
+    placeholder,
+    helperText,
+    startAdornmentText,
+    endAdornmentText,
+    startIcon: StartIcon,
+    endIcon: EndIcon,
+    ...other
+  } = props
+
+  /**
+   * Since Firefox doesn't support the :has() pseudo selector,
+   * manually track the focused state and use classes for focus, required, and disabled
+   */
+  const [isFocused, handleFocus, handleBlur] = useFocusState(
+    onFocusProp,
+    onBlurProp
+  )
+
+  // For focus behavior and accessiblity, <label> needs to have a htmlFor={} provided to an id matching the input
+  const backupId = useId()
```

Comment:
Aren't we wrapping the label around the input? I don't think this is necessary

Reply:
Same answer as above. It's an accessibility best practice